File: src/main/java/io/zeitwert/fm/account/model/impl/ObjAccountPersistenceProvider.java
package io.zeitwert.fm.account.model.impl;

import static io.zeitwert.ddd.util.Check.requireThis;

import java.math.BigDecimal;

import org.jooq.DSLContext;
import org.jooq.exception.NoDataFoundException;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.DependsOn;

import io.zeitwert.ddd.persistence.jooq.AggregateState;
import io.zeitwert.ddd.persistence.jooq.base.ObjExtnPersistenceProviderBase;
import io.zeitwert.fm.account.model.ObjAccount;
import io.zeitwert.fm.account.model.db.Tables;
import io.zeitwert.fm.account.model.db.tables.records.ObjAccountRecord;

@Configuration("accountPersistenceProvider")
@DependsOn("codePartListTypeEnum")
public class ObjAccountPersistenceProvider extends ObjExtnPersistenceProviderBase<ObjAccount> {

	public ObjAccountPersistenceProvider(DSLContext dslContext) {
		super(ObjAccount.class, dslContext);
		this.mapField("name", AggregateState.EXTN, "name", String.class);
		this.mapField("description", AggregateState.EXTN, "description", String.class);
		this.mapField("accountType", AggregateState.EXTN, "account_type_id", String.class);
		this.mapField("clientSegment", AggregateState.EXTN, "client_segment_id", String.class);
		this.mapField("referenceCurrency", AggregateState.EXTN, "reference_currency_id", String.class);
		this.mapField("inflationRate", AggregateState.EXTN, "inflation_rate", BigDecimal.class);
		this.mapField("logoImage", AggregateState.EXTN, "logo_img_id", Integer.class);
		this.mapField("mainContact", AggregateState.EXTN, "main_contact_id", Integer.class);
	}

	@Override
	public ObjAccount doCreate() {
		return this.doCreate(this.dslContext().newRecord(Tables.OBJ_ACCOUNT));
	}

	@Override
	public ObjAccount doLoad(Integer objId) {
		requireThis(objId != null, "objId not null");
		ObjAccountRecord accountRecord = this.dslContext().fetchOne(Tables.OBJ_ACCOUNT,
				Tables.OBJ_ACCOUNT.OBJ_ID.eq(objId));
		if (accountRecord == null) {
			throw new NoDataFoundException(this.getClass().getSimpleName() + "[" + objId + "]");
		}
		return this.doLoad(objId, accountRecord);
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/account/persist/jooq/ObjAccountPersistenceProvider.kt
package io.zeitwert.fm.account.persist.jooq

import io.dddrive.core.property.model.BaseProperty
import io.dddrive.core.property.model.EnumProperty
import io.dddrive.core.property.model.ReferenceProperty
import io.zeitwert.dddrive.ddd.persist.jooq.JooqObjPersistenceProviderBase
import io.zeitwert.fm.account.model.ObjAccount
import io.zeitwert.fm.account.model.ObjAccountRepository
import io.zeitwert.fm.account.model.db.Tables
import io.zeitwert.fm.account.model.enums.CodeAccountType
import io.zeitwert.fm.account.model.enums.CodeClientSegment
import io.zeitwert.fm.account.model.enums.CodeCurrency
import io.zeitwert.fm.obj.model.db.Sequences
import org.jooq.DSLContext
import org.jooq.UpdatableRecord
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component("objAccountPersistenceProvider")
open class ObjAccountPersistenceProvider : JooqObjPersistenceProviderBase<ObjAccount>() {

	private lateinit var _dslContext: DSLContext
	private lateinit var _repository: ObjAccountRepository

	@Autowired
	fun setDslContext(dslContext: DSLContext) {
		this._dslContext = dslContext
	}

	@Autowired
	@Lazy
	fun setRepository(repository: ObjAccountRepository) {
		this._repository = repository
	}

	override fun dslContext(): DSLContext = _dslContext

	override fun getAggregateTypeId(): String = AGGREGATE_TYPE_ID

	override fun nextAggregateId(): Any =
		dslContext()
			.nextval(Sequences.OBJ_ID_SEQ)
			.toInt()

	override fun fromAggregate(aggregate: ObjAccount): UpdatableRecord<*> = createObjRecord(aggregate)

	@Suppress("UNCHECKED_CAST")
	override fun storeExtension(aggregate: ObjAccount) {
		val objId = aggregate.id as Int

		val existingRecord = dslContext().fetchOne(
			Tables.OBJ_ACCOUNT,
			Tables.OBJ_ACCOUNT.OBJ_ID.eq(objId),
		)

		val record = existingRecord ?: dslContext().newRecord(Tables.OBJ_ACCOUNT)

		record.objId = objId
		record.tenantId = aggregate.tenantId as? Int
		record.accountId = (aggregate.getProperty("accountId") as? BaseProperty<Int?>)?.value
		record.name = (aggregate.getProperty("name") as? BaseProperty<String?>)?.value
		record.description = (aggregate.getProperty("description") as? BaseProperty<String?>)?.value
		record.accountTypeId = (aggregate.getProperty("accountType") as? EnumProperty<CodeAccountType>)?.value?.id
		record.clientSegmentId = (aggregate.getProperty("clientSegment") as? EnumProperty<CodeClientSegment>)?.value?.id
		record.referenceCurrencyId = (aggregate.getProperty("referenceCurrency") as? EnumProperty<CodeCurrency>)?.value?.id
		record.inflationRate = (aggregate.getProperty("inflationRate") as? BaseProperty<BigDecimal?>)?.value
		record.discountRate = (aggregate.getProperty("discountRate") as? BaseProperty<BigDecimal?>)?.value
		record.logoImgId = (aggregate.getProperty("logoImage") as? ReferenceProperty<*>)?.id as? Int
		record.mainContactId = (aggregate.getProperty("mainContact") as? ReferenceProperty<*>)?.id as? Int

		if (existingRecord != null) {
			record.update()
		} else {
			record.insert()
		}
	}

	@Suppress("UNCHECKED_CAST")
	override fun loadExtension(
		aggregate: ObjAccount,
		objId: Int?,
	) {
		if (objId == null) return

		val record = dslContext().fetchOne(
			Tables.OBJ_ACCOUNT,
			Tables.OBJ_ACCOUNT.OBJ_ID.eq(objId),
		) ?: return

		(aggregate.getProperty("accountId") as? BaseProperty<Int?>)?.value = record.accountId
		(aggregate.getProperty("name") as? BaseProperty<String?>)?.value = record.name
		(aggregate.getProperty("description") as? BaseProperty<String?>)?.value = record.description

		record.accountTypeId?.let { accountTypeId ->
			(aggregate.getProperty("accountType") as? EnumProperty<CodeAccountType>)?.value =
				CodeAccountType.getAccountType(accountTypeId)
		}

		record.clientSegmentId?.let { clientSegmentId ->
			(aggregate.getProperty("clientSegment") as? EnumProperty<CodeClientSegment>)?.value =
				CodeClientSegment.getClientSegment(clientSegmentId)
		}

		record.referenceCurrencyId?.let { currencyId ->
			(aggregate.getProperty("referenceCurrency") as? EnumProperty<CodeCurrency>)?.value =
				CodeCurrency.getCurrency(currencyId)
		}

		(aggregate.getProperty("inflationRate") as? BaseProperty<BigDecimal?>)?.value = record.inflationRate
		(aggregate.getProperty("discountRate") as? BaseProperty<BigDecimal?>)?.value = record.discountRate

		(aggregate.getProperty("logoImage") as? ReferenceProperty<*>)?.let { prop ->
			@Suppress("UNCHECKED_CAST")
			(prop as ReferenceProperty<Any>).id = record.logoImgId
		}

		(aggregate.getProperty("mainContact") as? ReferenceProperty<*>)?.let { prop ->
			@Suppress("UNCHECKED_CAST")
			(prop as ReferenceProperty<Any>).id = record.mainContactId
		}
	}

	companion object {

		private const val AGGREGATE_TYPE_ID = "obj_account"
	}
}
